### PR TITLE
improvement currentTimeMillis calc, using System.nanoTime() to measur…

### PR DESCRIPTION
**Pull request by wuwen5**
improvement currentTimeMillis calc, using System.nanoTime() to measure elapsed time.

### CODE REVIEW

File: core/src/main/java/com/alibaba/nacos/core/distributed/id/SnowFlowerIdGenerator.java [hunk 1]
```diff
@@ -72,13 +72,19 @@
...
+
+    //CLOCK_REALTIME
+    private final long startWallTime = System.currentTimeMillis();
+
+    //CLOCK_MONOTONIC
+    private final long monotonicStartTime = System.nanoTime();
+
     private long workerId;
-    
+
```
Comment (KomachiSion) [inline core/src/main/java/com/alibaba/nacos/core/distributed/id/SnowFlowerIdGenerator.java:83]: Do not change indent , please use nacos code style to reformat codes.

**Review by KomachiSion — CHANGES_REQUESTED** [review]

**Review by KomachiSion — COMMENTED** [review]
Have you tested how much performance this modification can improve?

**Comment by wuwen5** [comment]
It mainly not affected by  clock moving backwards, Using ```System.currentTimeMillis()``` to measure elapsed time can give incorrect results, ```currentTimeMillis()``` is wall time, backwards is possible, ```nanoTime()``` is monotonic time It's impossible to go backwards.

**Comment by KomachiSion** [comment]
But I see there is much cal for time, have you test the performance effect?

**Comment by wuwen5** [comment]
```
# Mac OS X: 10.15.7
# JMH version: 1.32
# VM version: JDK 1.8.0_181, Java HotSpot(TM) 64-Bit Server VM, 25.181-b13
Benchmark                     Mode  Cnt       Score        Error  Units
Test.newImplUseNanoTimeCalc  thrpt    5  610940.234 ±  54002.164  ops/s
Test.originally              thrpt    5  535988.141 ± 105961.666  ops/s


# Linux Red Hat 4.8.5-28:3.10.0-862.el7.x86_64
# JMH version: 1.32
# VM version: JDK 1.8.0_291, Java HotSpot(TM) 64-Bit Server VM, 25.291-b10
Benchmark                     Mode  Cnt       Score        Error  Units
Test.newImplUseNanoTimeCalc  thrpt    5  557613.023 ± 184723.957  ops/s
Test.originally              thrpt    5  515881.617 ± 195077.400  ops/s
```

**Comment by wuwen5** [comment]
BTW:if adding log judgment
```java
if (logger.isDebugEnabled()) {
    logger.debug("{}-{}-{}", (new SimpleDateFormat(DATETIME_PATTERN)).format(new Date(this.lastTime)),
            workerId, this.sequence);
}
```


```
Benchmark                     Mode  Cnt        Score        Error  Units
Test.newImplUseNanoTimeCalc  thrpt    5  4070147.689 ± 102391.074  ops/s
Test.originally              thrpt    5  3903584.317 ± 590721.088  ops/s
```

**Comment by KomachiSion** [comment]
It seems no much improve. So just to fix the possible clock backward？

**Review by KomachiSion — COMMENTED** [review]
I think this logic should exist.

Or use wait to next mills to replace

**Comment by wuwen5** [comment]
> It seems no much improve. So just to fix the possible clock backward？

Yes

File: core/src/main/java/com/alibaba/nacos/core/distributed/id/SnowFlowerIdGenerator.java [hunk 2]
```diff
@@ -111,10 +117,7 @@ public long currentId() {
     
     @Override
     public synchronized long nextId() {
-        long currentMillis = System.currentTimeMillis();
-        Preconditions.checkState(this.lastTime <= currentMillis,
```
[the remark this answers is not in the document]
Comment (wuwen5) [reply core/src/main/java/com/alibaba/nacos/core/distributed/id/SnowFlowerIdGenerator.java:115]: the clock backward has been fixed, so this logic will never false

**Comment by KomachiSion** [comment]
OK, Can you add some unit test to test your change?

**Comment by wuwen5** [comment]
@KomachiSion   the source file ```SnowFlowerIdGenerator_ITCase.java``` wrong uses tabs, so changed indent

**Comment by KomachiSion** [comment]
> @KomachiSion the source file `SnowFlowerIdGenerator_ITCase.java` wrong uses tabs, so changed indent

Please do not change any ITCase, Please add the unit test not IT.

**Review by KomachiSion — APPROVED** [review]

**Pull request merged**